### PR TITLE
Measure Code Coverage

### PR DESCRIPTION
**Pull request by swissspidy**
While working on #2408 I realized there was currently no easy way to measyre PHP code coverage for the plugin.

With this PR, code coverage analysis is automatically run on Travis CI and then uploaded to [Coveralls](https://coveralls.io/). This is the only service that wp-dev-lib supports out of the box. I've previously only used Codecov, but I guess they are pretty much the same.

Also, this PR disables Xdebug when no coverage analysis is needed, for a nice little extra boost.

Wanna get some code coverage analysis _today_? Fear not! It's as easy as following these steps:

1. Make sure Xdebug is installed and active on your computer. Otherwise use `pecl install xdebug`
2. Run `phpunit --coverage-clover build/logs/clover.xml`. This might take a while.
3. In IntellJ, open Analyze -> Show Code Coverage Data and select the newly created XML file.
4. Get visualization like this:

Screenshot 2019-05-27 at 17 42 39

### CODE REVIEW

**Comment by swissspidy** [comment]
After disabling Xdebug when not needed, the Travis jobs are now super fast: https://travis-ci.org/ampproject/amp-wp/builds/537875524 ⚡️

**Comment by westonruter** [comment]
@swissspidy The code coverage job is failing: https://travis-ci.org/ampproject/amp-wp/jobs/537889379

image

```
Read environment variables
Dump submitting json file: /tmp/wordpress/src/wp-content/plugins/amp/build/logs/coveralls-upload.json
File size: 2,012.67 kB
Submitting to https://coveralls.io/api/v1/jobs
Client error occurred. status: 422 Unprocessable Entity
Couldn't find a repository matching this job.
elapsed time: 1.488 sec memory: 14.00 MB
```

**Comment by swissspidy** [comment]
@westonruter That is to be expected, as the the project hasn’t been set up on Coveralls yet. See todo above:

> Verify whether Coveralls can actually be added (GitHub permissions)

I don‘t have the permissions to add the Coveralls GitHub app to this repository. Without that, Coveralls cannot set things up.

**Comment by swissspidy** [comment]
Looks like the main repo uses Codecov, so I am inclined to use the same: https://codecov.io/gh/ampproject/amphtml/

**Comment by swissspidy** [comment]
Just pushed some config changes that should make it work with Codecov. Nice bonus is that Codecov apparently makes it super easy to upload reports for both PHP and JS coverage.

I also requested permissions in order to get access to this repo with Codecov:

Screenshot 2019-05-29 at 11 12 17

Probably necessary because I am only a collaborator on this repo. Maybe you can directly add it without problems, @westonruter.

**Comment by swissspidy** [comment]
The last build failed because of PHPUnit errors:

```
PHP Fatal error:  Uncaught Error: Class 'AMP_Base_Sanitizer' not found in /tmp/wordpress/src/wp-content/plugins/amp/tests/test-class-amp-theme-support.php:1907
Stack trace:
#0 phar:///home/travis/phpunit-bin/5.7/phpunit/phpunit/Util/Fileloader.php(52): include_once()
#1 phar:///home/travis/phpunit-bin/5.7/phpunit/phpunit/Util/Fileloader.php(36): PHPUnit_Util_Fileloader::load('/tmp/wordpress/...')
#2 phar:///home/travis/phpunit-bin/5.7/phpunit/phpunit/Framework/TestSuite.php(316): PHPUnit_Util_Fileloader::checkAndLoad('/tmp/wordpress/...')
#3 phar:///home/travis/phpunit-bin/5.7/phpunit/phpunit/Framework/TestSuite.php(393): PHPUnit_Framework_TestSuite->addTestFile('/tmp/wordpress/...')
#4 phar:///home/travis/phpunit-bin/5.7/phpunit/phpunit/Util/Configuration.php(946): PHPUnit_Framework_TestSuite->addTestFiles(Array)
#5 phar:///home/travis/phpunit-bin/5.7/phpunit/phpunit/Util/Configuration.php(843): PHPUnit_Util_Configuration->getTestSuite(Object(DOMElement), NULL)
#6 phar:///home/travis/phpunit-bin/5.7/phpunit/phpunit/TextUI/Com in /tmp/wordpress/src/wp-content/plugins/amp/tests/test-class-amp-theme-support.php on line 1907
Fatal error: Uncaught Error: Class 'AMP_Base_Sanitizer' not found in /tmp/wordpress/src/wp-content/plugins/amp/tests/test-class-amp-theme-support.php on line 1907
Error: Class 'AMP_Base_Sanitizer' not found in /tmp/wordpress/src/wp-content/plugins/amp/tests/test-class-amp-theme-support.php on line 1907
```

No idea why this happens. Seems like some autoloader issue, but I couldn't reproduce it locally.

See https://travis-ci.org/ampproject/amp-wp/jobs/538685863#L685-L710

**Comment by westonruter** [comment]
> PHP Fatal error:  Uncaught Error: Class 'AMP_Base_Sanitizer' not found in /tmp/wordpress/src/wp-content/plugins/amp/tests/test-class-amp-theme-support.php:1907

@swissspidy This happens when `amp.php` short-curcuits due to the plugin not having been built, as I see you have discovered!

**Comment by westonruter** [comment]
@swissspidy I've put in the request to get the GitHub app installed.

I suppose we need to update wp-dev-lib to enable generation of the badge in `readme.md`.

**Comment by swissspidy** [comment]
Yeah wp-dev-lib doesn‘t support codecov yet. Not urgent, as it‘s mostly important for ourselves.

Let‘s hope the request gets through soon. Maybe we can ping some people directly?

File: .travis.yml [hunk 1]
```diff
@@ -110,6 +115,20 @@ jobs:
...
+      php: "7.3"
+      env: WP_VERSION=trunk  DEV_LIB_ONLY=phpunit                               INSTALL_PWA_PLUGIN=1
+      before_script:
+        - echo "Running unit tests with code coverage..."
+      script:
+        - npm run build:js
+        - source "$DEV_LIB_PATH/travis.script.sh"
+        - bash <(curl -s https://codecov.io/bash) -cF php
+        - npm run test:js
+        - bash <(curl -s https://codecov.io/bash) -cF javascript
```
Comment (rsimha) [inline .travis.yml]: If this doesn't work, I'd suggest trying https://www.npmjs.com/package/codecov.

See [here](https://github.com/ampproject/amphtml/blob/master/build-system/tasks/codecov-upload.js#L27-L47) for how it's used by `ampproject/amphtml`.

**Comment by swissspidy** [comment]
The issue with PHP coverage is a bit odd: https://travis-ci.org/ampproject/amp-wp/jobs/539185192#L979-L1007

PHPUnit says a coverage report was built, but Codecov can't find anything at `build/logs/clover.xml`.

**Comment by westonruter** [comment]
> PHPUnit says a coverage report was built, but Codecov can't find anything at `build/logs/clover.xml`.

I believe the problem is that the [`travis.after_script.sh`](https://github.com/xwp/wp-dev-lib/blob/master/scripts/travis.after_script.sh) hasn't been run. So that needs to be sourced as well, before running `bash <(curl -s https://codecov.io/bash) -cF php -f ...`

**Comment by swissspidy** [comment]
Hmm that just uploads the report to Coveralls, no? We already (try to) upload it to Codecov in the script step, so that doesn‘t really apply to this setup.

The actual code coverage analysis is done by PHPUnit while running the tests.

File: amp.php [hunk 2]
```diff
@@ -185,15 +185,24 @@ function _amp_show_load_errors_admin_notice() {
 // Abort if dependencies are not satisfied.
 if ( ! empty( $_amp_load_errors->errors ) ) {
 	add_action( 'admin_notices', '_amp_show_load_errors_admin_notice' );
-	if ( defined( 'WP_CLI' ) && WP_CLI ) {
+
+	if ( ( defined( 'WP_CLI' ) && WP_CLI ) || 'true' === getenv( 'CI' ) ) {
```
Comment (swissspidy) [inline amp.php]: This output is useful when running the tests locally as well.

**Comment by swissspidy** [comment]
**Finally! It works! 🥳**

Check out the newest code coverage report here: https://codecov.io/gh/ampproject/amp-wp/tree/140e6b3abdd1c86cf10cdbee8c2fa3e19b85824d

It combines both PHP and JS coverage, which is nice.

**Comment by amedina** [comment]
Excellent!

Em sex, 31 de mai de 2019 às 12:07, Pascal Birchler <
notifications@github.com> escreveu:

**Pull request merged**